File: Benchmarking/Dart/matinv.dart
import 'dart:io';

void main() {
  List<List<double>> matrix = [];
  List<List<double>> inverse = [];

  // Read matrix elements from file
  readMatrix('../random_matrix.txt', matrix);

  // Resize the inverse matrix to the size of the original matrix
  int n = matrix.length;
  inverse = List.generate(n, (_) => List.filled(n, 0.0));

  // Compute the inverse matrix
  gaussJordanInverse(matrix, inverse);

  // Write the inverse matrix to a file
  writeMatrixToFile('./inverse_matrix.txt', inverse);
}

void readMatrix(String filename, List<List<double>> matrix) {
  File file = File(filename);
  if (!file.existsSync()) {
    throw Exception("Error opening file");
  }

  List<String> lines = file.readAsLinesSync();
  int n = int.parse(lines[0]);

  matrix.clear();
  for (int i = 0; i < n; i++) {
    List<double> row = lines[i + 1].split(' ').map((e) => double.parse(e)).toList();
    matrix.add(row);
  }
}

void writeMatrixToFile(String filename, List<List<double>> matrix) {
  File file = File(filename);
  IOSink sink = file.openWrite();

  for (List<double> row in matrix) {
    sink.writeln(row.map((e) => e.toString()).join(' '));
  }

  sink.close();
}

void gaussJordanInverse(List<List<double>> matrix, List<List<double>> inverse) {
  int n = matrix.length;
  List<List<double>> augmented = List.generate(n, (_) => List.filled(2 * n, 0.0));

  // Create an augmented matrix with the identity matrix on the right
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      augmented[i][j] = matrix[i][j];
      augmented[i][j + n] = (i == j) ? 1.0 : 0.0;
    }
  }

  // Perform Gauss-Jordan elimination
  for (int i = 0; i < n; i++) {
    double pivot = augmented[i][i];
    if (pivot == 0) {
      int swapRow = -1;
      for (int j = i + 1; j < n; j++) {
        if (augmented[j][i] != 0) {
          swapRow = j;
          break;
        }
      }
      if (swapRow == -1) {
        throw Exception("Matrix is singular and cannot be inverted.");
      }

      // Swap rows
      List<double> temp = augmented[i];
      augmented[i] = augmented[swapRow];
      augmented[swapRow] = temp;

      pivot = augmented[i][i];
    }

    // Normalize the pivot row
    for (int j = 0; j < 2 * n; j++) {
      augmented[i][j] /= pivot;
    }

    // Eliminate the current column in all other rows
    for (int j = 0; j < n; j++) {
      if (j != i) {
        double factor = augmented[j][i];
        for (int k = 0; k < 2 * n; k++) {
          augmented[j][k] -= factor * augmented[i][k];
        }
      }
    }
  }

  // Extract the inverse matrix from the augmented matrix
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      inverse[i][j] = augmented[i][j + n];
    }
  }
}

File: Benchmarking/Julia/matinv.jl
# Function to read the matrix from a file
function read_matrix(filename::String)
    matrix = []
    open(filename, "r") do file
        n = parse(Int, readline(file))  # Read the size of the matrix
        matrix = [parse.(Float64, split(readline(file))) for _ in 1:n]
    end
    return matrix
end

# Function to write matrix to file
function write_matrix_to_file(filename::String, matrix::Vector{Vector{Float64}})
    open(filename, "w") do file
        for row in matrix
            println(file, join(map(x -> string(x), row), " "))
        end
    end
end

# Gauss-Jordan inversion
function gauss_jordan_inverse(matrix::Vector{Vector{Float64}})
    n = length(matrix)
    augmented = [hcat(matrix[i], [i == j ? 1.0 : 0.0 for j in 1:n]) for i in 1:n]

    # Perform Gauss-Jordan elimination
    for i in 1:n
        pivot = augmented[i][i]
        if pivot == 0.0
            swap_row = -1
            for j in i+1:n
                if augmented[j][i] != 0.0
                    swap_row = j
                    break
                end
            end
            if swap_row == -1
                throw(ArgumentError("Matrix is singular and cannot be inverted."))
            end

            # Swap rows
            augmented[i], augmented[swap_row] = augmented[swap_row], augmented[i]
            pivot = augmented[i][i]
        end

        # Normalize the pivot row
        for j in 1:2n
            augmented[i][j] /= pivot
        end

        # Eliminate the current column in all other rows
        for j in 1:n
            if j != i
                factor = augmented[j][i]
                for k in 1:2n
                    augmented[j][k] -= factor * augmented[i][k]
                end
            end
        end
    end

    # Extract the inverse matrix from the augmented matrix
    inverse = [augmented[i][n+1:end] for i in 1:n]
    return inverse
end


# Main function
function main()
    # Read the matrix from a file
    matrix = read_matrix("../random_matrix.txt")

    # Compute the inverse
    inverse = gauss_jordan_inverse(matrix)

    # Write the inverse matrix to a file
    write_matrix_to_file("inverse_matrix.txt", inverse)
end

# Run the main function
main()